man/logsurfer.1.in
.\"
.\" written by W. Ley, U. Ellermann, and K. Thompson
.\"
.\" @(#) $Header: logsurfer.1,v 
.TH LOGSURFER 1  "Thu Oct 19 1995"
.SH NAME
logsurfer \- process logfiles and perform certain actions
.SH SYNOPSIS
.B logsurfer
[
.B \-c configfile
] [
.B \-d dumpfile
] [
.B \-f
] [
.B \-F
] [
.B \-l start_linenum
] [
.B \-p pidfile
] [
.B \-r start_regex
] [
.B \-s
] [
.B \-t
] [
.B \-e
] [
.B filename
]
.SH DESCRIPTION
.LP
\fILogsurfer\fP is a program that works on single textlines from a
logfile (or from standard input). It compares each line against regular
expressions of a set of "rules" and performs certain actions that are
defined for this match.
.LP
The \fIlogsurfer\fP uses "rules" to define actions for messages (single
lines). Messages may be collected in "contexts" which are stored in memory.
Those contexts have their own default actions (e.g. for timeouts). At
starttime you have to define a set of rules in the configuration file.
While processing logmessages the logsurfer is able to dynamically add or
delete rules, create or destroy contexts, invoke external programs for
alerting people or to perform further automated actions etc.
.LP
A \f4rule\f1 starts with a set of regular expressions. The messageline
that is being checked has to match against the first regular expression
but must not
match the second expression. The string "-" can be used as the second
regex if you don't need this "match not" expression. If the messageline
meets these requirements the associated action is taken. Currently the
following actions are allowed: ignore this message, execute an external
program, pass this messageline as standard input to an external program,
create or delete a context, generate a "report" or dynamically add a new
rule. In addition to these actions rules may have a configuration to
delete themself if another message (again specified by two regular
expressions) is being processed by the logsurfer or if an optional
timeout has occured. Matching of messagelines against rules is usually done
until the first match occurs. This can be circumvented by specifying a
"continue" keyword in the rule. In this case the program will continue
to search for another match in the rule definitions.
.LP
A \f4context\f1 is a set of messagelines that have been read by the
logsurfer. Again the messagelines are specified by two regular expressions.
Contexts can be used in "reports" to present a collection of messages
associated with a specific action (e.g. all messagelines from one ftp
session). Every context has a default action which is launched if either
a specified timeout has occured or a configured limit (of linenumbers) has
been reached. Default actions may be: ignore (drop) this context, execute
an external program, feed all collected logmessages as standard input to
an external program or generate a "report".
.LP
A \f4report\f1 summarizes a number of contexts and pipes the result
into an external program.
.LP
The normal procedure for each logline is: Store this line in all matching
contexts (as defined by their regular expressions), if the self-destroy
regex matches, too, then the default action of the context is executed and the
context is deleted, find the first matching rule and execute the given action
(continue rule matching if the rule has the keyword "continue" --- see
logsurfer.conf(4) manpage), check for the stop-pattern of the rule and in case
of a match delete the rule, check for timeouts of rules and delete the
affected rules, check for context timeouts and execute their default action.
.LP
Errors (like broken configuration statements or memory problems) are 
reported on standard error. The \fIsyslog\fP mechanism is currently not
being used because this might result in a loop if the logsurfer is processing
those syslog-messages. It is suggested to start the logsurfer and redirect
the output (incl. error output) to another file. If necessary you might
want to start a second logsurfer that is watching the output of the first
logsurfer.
.LP
If the program exits the following steps are executed:
.TP 6
1)
The status of all active rules and contexts is dumped to the dumpfile.
.TP
2)
If contexts have timeout values (other than 0) then a timeout is send to
those functions. This will result in launching the default actions of the
affected contexts.
.TP
3)
Rules and contexts are deleted from memory (cleanup). This is only important
for debugging purposes and should have no effect.
.TP
4)
If there are still external programs running the logsurfer will wait up to
20 seconds to catch them.
.TP
5)
The logsurfer exits with a return value of zero (0).
.LP
.LP
.SH OPTIONS
.TP
.B \-c configfile
Specify another configuration file for the logsurfer. The default is
"@CONFFILE@".
.TP
.B \-d dumpfile
Specify another dumpfile for the logsurfer. The logsurfer writes the internal
information about rules and contexts to this file if the program exists or a
special signal was sent to the process (see NOTES). The default is
"@DUMPFILE@".
.TP
.B \-D
Daemonize. Fork and run the logsurfer process in the background.
Any output sent to stdin, stdout and stderr are redirected to the
bitbucket.
.TP
.B \-f
Normally the logsurfer will process the input once and exit at the end
(after processing all the input lines). By using the "-f" option the
program will enter a special "follow" mode if the end of file has been
reached. In this case the program will sleep for one second before trying
to read a new line. If a new line has been added to the end of the input,
then this line is processed. Otherwise the program will sleep again for
one second. This mode is similar to the "-f" option of the tail(1)
command.
.TP
.B \-F
Acts like the "-f" option, but follows the logfile across rotations by
checking if the inode of the file has changed and re-opening the log.
.TP
.B \-l start_linenum
Start logfile processing at the given linenumber.
.TP
.B \-p pidfile
Write the processid of the logsurfer process to the given filename. This
can be used to simplify sending signals to the logsurfer program (for
example after shifting logfiles).
.TP
.B \-r start_regex
Start logfile processing with the first line that matches the given regular
expression.
.TP
.B \-s
Normally the logsurfer prints some information to stderr if the
program exists. If you specify the "-s" (silent) option, then those
messages are not printed.
.TP
.B \-t
When logsurfer exits, the -t option will instruct it to explicitly expire
all contexts which have either a relative or absolute timeout configured.
Without the -t option, contexts will not be timed-out and will be removed
silently without triggering their actions.
.TP
.B \-e
When Logsurfer starts reading the log file this option will instruct it to
seek to the end of the file before scanning starts. Without this option,
Logsurfer will start scanning at the beginning of the file.
.SH "FILES"
.PD 0
.TP 32
.B @CONFFILE@
default configuration file
.TP
.B @DUMPFILE@
dump of the rules and contexts
.PD
.SH "SEE ALSO"
.BR logsurfer.conf(4),
.BR swatch(8)
.SH NOTES
This program may consume a lot of memory. Resource control should be
done by the operating system (for example by setting limits). The program
was designed to fall back to a defined state if some resources are not
available.
.LP
Please do not run the logsurfer with a privileged userid (like root). You
usually do not need the extra privileges. If your logfiles are not
world-readable change them to be group readable and start the logsurfer
with a non-privileged account that is a member of that group. Keep in
mind, that the logsurfer might execute external programs (depending on
your configuration file) which are not designed to run with higher
privileges. The logsurfer program will print a warning message if it's
started with root privilege (this feature can be disabled at compile time).
.LP
Be careful with external programs. The logsurfer is able to process any
input (for example arbitrary long lines) but some other programs may
crash or produce other inpredictable results if feeded with long lines.
.LP
The following signals have the specified effect when sent to the logsurfer
program using the
.BR kill(1)
command.
.TP 10
.SM SIGHUP
Reopens the logfile (to be analyzed). Processing starts with the first line
in this file. This can be used after "shifting" logfiles and creating new
logfiles.
.TP
.SM SIGUSR1
Dump the internal state of all active rules and the collected contexts to
the dumpfile "@DUMPFILE@".
.TP
.SM SIGTERM
Dumps the state to the dumpfile (see above) and exits.
.SH BUGS
Quoting of regular expressions isn't that easy. If you use variables
($0 - $9) in dynamic rules, then you have to keep in mind, that certain
characters (matching your regular expresion for that variable) may
again contain special chars, that are interpreted as regex-parts.
Example: If you use one regular expresion to match a hostname and you
are using this again as part of a regular expression, then the dots
"." separating parts of the domainname will match any char (not only
the dot).
.LP
If the logsurfer has been compiled with the "sendmail flush hack" and the
logsurfer exists, then there might still be some mails in the local
sendmail queue (not flushed). Delivery of such mail maybe delayed until
the next time sendmail flushes the queue.



ChangeLog
Version 1.8 (logsurfer)
	Kerry Thompson kerry@crypt.gen.nz:
	+ fixed double free() in exec.c/prepare_exec(), thanks to reports
	  from Gregor Kopf of Recurity Labs and Jan Kohlrausch of DFN_CERT
	+ centralised version number in Makefile.in, added -DVERSION=...
	+ updated README & DISCLAIMER files removing DFN-CERT copyright
	+ now just Logsurfer, not Logsurfer+ any more
	+ -D command line option for daemon mode. Warning: closes stdin,
	  stdout, & stderr, therefore no error messages
	+ -F command line option to auto re-open log file
	+ dropped the config-examples directory, added note in README
	  to go to http://www.crypt.gen.nz/logsurfer/#configs

Version 1.7 (logsurfer+)
	Kerry Thompson kerry@crypt.gen.nz:
	+ added -e option to begin processing from the current end
	  of the input log file ( normally used with -f )
	+ put double-quotes around regex expressions in dump file
	+ if the context argument to a pipe or report action is "-"
	  then the current context contents are piped into the command
	  this should shorten most context definitions
        + added new action "echo" which simply echo's the output on
          stdout, or to a file with optional >file & >>file 1st arg
	+ added a macro construct in context action fields, if "$lines"
	  exists in a context action (such as a command line) it will
	  be substituted by the number of lines in the context
	+ added syslog action to send a message into syslog
	  the first argument must be facility:level, the second arg
	  is a string to send to syslog

Version 1.6b (logsurfer+)
	+ fixed -t option in getopts()

Version 1.6a (logsurfer+)
	Kerry Thompson kerry@crypt.gen.nz:
	+ corrected off-by-one in min_lines checking
	+ added min_lines context arg to logsurfer.conf man page

Version 1.6 (logsurfer+)
	Kerry Thompson kerry@crypt.gen.nz:
	+ added an optional parameter at the end of context definition
	  (just before action) specifying the minimum number of lines
	  collected which needs to be satisfied before performing action
	+ added -t option to explicity timeout contexts when exiting
	  default=off so contexts don't all trigger when we shut down
	+ only add lines to a context if the context has an action of
	  'pipe' or 'report'. In other words, don't store lines in
	  memory which won't ever be used. The number of matching
	  lines in the context is still incremented. This allows contexts
	  to be created which can notify if we _don't_ see an event, such
	  as regular "syslog pings" from hosts.

Version 1.5b
	+ Corrected off-by-one heap overflow and uninitialized pointer

Version 1.5a
	+ Corrected minor bug in context timeouts

Version 1.5
	+ Added y2k disclaimer
	+ Corrected authorship and copyright information
	+ Released

Version 1.5beta2
	+ Patch for context linelimit added
	+ Changed Logsurfer URL to http://www.cert.dfn.de/eng/logsurf/

Version 1.5beta
	+ Changed default dumpfile to /dev/null to avoid security problems.
	  You can specify the default location for the logsurfer.dump file
	  in the Makefile (see DEF_DUMPFILE). However - be aware that if using
	  a public writable directory someone can crate a sym-link for that
	  name and you'll write to another file. So I would suggest to use
	  the "-d" option while starting logsurfer to specify a dumpfilename.
	+ Added -DWARN_ROOT to print a warning if logsurfer is running as root.
	+ Added generic GNU "INSTALL" file
	+ Corrected version number in Makefile

Version 1.41
	+ deleted two #error defines which caused problems on non
	  ansi-compilers

Version 1.4
	+ fixed serious bug: rule processing stopped if a match_regex was
	  found (even if the not_match_regex dropped this rule)
	+ setup www-page http://www.cert.dfn.de/eng/team/wl/logsurfer/

Version 1.3
	+ removed #error directive (this is only available on ansi
	  compilers)
	+ added output of logsurfer version number in the usage
	  output

Version 1.2
	+ updated manpage of logsurfer.1 to include -f and the new
	  -p option
	+ added -p option to write pid to a file (for easier restarts)
	+ changed line handling to be independent from line length
	  (no fixed buffers for processing stuff)
	+ added --with-etcdir option for configure to specify the
	  default location of the logsurfer.conf file
	+ new "contrib" directory to collect helpful tools (added a script
	  "start-mail" and a tool "prtime" to start the collection)
	+ added a check if "const" is defined or not
	+ sendmail flushing has been improved (but is still not completly
	  reliable - it is only implemented as an additional help for
	  flushing the sendmail queue beside the time sendmail does it
	  by itself)
	+ -DSENDMAIL_FLUSH is now not longer default
	+ LDFLAGS in Makefile.in corrected
	+ trying to report linenumber on configuration errors
	+ the behaviour or fgets() at EOF is not portable - now using read()
	+ all header defines also for non ANSI C compilers available
	+ added sys/resource.h include because AIX header files are broken
	+ added srcdir and VPATH definition to Makefiles
	+ changed all #include "../config.h" to #include <config.h>
	+ added Makefile.in in man/ and src/ dir
	+ added PostScript version of regex documentation to package
	+ merged configure.in from the old regex dir to the new global one
	+ removed regex-0.12 dir and included regex.c and regex.h in the
	  logsurfer source dir

Version 1.1
	+ splited the code into several peaces
	+ wrote a draft manual page



src/logsurfer.c
/*
 * logsurfer
 *
 * Authors: Wolfgang Ley, Uwe Ellermann, Kerry Thompson
 */


/* config header if available */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif


/* system includes */

#include <stdio.h>

#if HAVE_STDLIB_H || STDC_HEADERS
#include <stdlib.h>
#endif

/* ugly, but portable */ 
#if HAVE_STRING_H || STDC_HEADERS
#include <string.h>
#else
#if HAVE_STRINGS_H
#include <strings.h>
#else
#if HAVE_STRCHR
char *strchr();
#else
#if HAVE_INDEX
#define strchr index
char *strchr();
#endif
#endif
#endif
#endif

#include <ctype.h>

#if HAVE_LIMITS_H
#include <limits.h>
#endif

#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <paths.h>

#if TIME_WITH_SYS_TIME
#include <sys/time.h>
#include <time.h>
#else
#if HAVE_SYS_TIME_H
#include <sys/time.h>
#else
#include <time.h>
#endif
#endif

#if HAVE_SYS_RESOURCE_H
#include <sys/resource.h>
#endif

#if HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif

#if HAVE_UNISTD_H
#include <unistd.h>
#endif

#if HAVE_MALLOC_H
#include <malloc.h>
#endif

#include <sys/errno.h>
#include <fcntl.h>

/* local includes */

#include "logsurfer.h"
#include "globals.h"
#include "regex.h"
#include "str_util.h"
#include "context.h"
#include "rule.h"
#include "exec.h"
#include "exit.h"
#include "readline.h"
#include "readcf.h"


/*
 * the global variables
 */

#ifndef MAXPATHLEN
#define MAXPATHLEN 1024
#endif

struct rule	*all_rules;	/* all active rules		*/
struct rule	*all_rules_end;
struct context	*all_contexts;	/* all active contexts		*/
struct context	*all_contexts_end;
long		next_rule_timeout;	/* next rule timeout	*/
long		next_context_timeout;	/* next context timeout	*/

char		logfile_name[MAXPATHLEN];	/* name of the logfile	*/
FILE		*logfile;		/* input logfile	*/
long		logline_num;	/* linenumer of the logline	*/
char		*logline;	/* the actual logline		*/
struct context_line *logline_context;/* the context_line ptr	*/
time_t		current_time;	/* current time information	*/

char		dumpfile_name[MAXPATHLEN];	/* name of the dumpfile */
char		pidfile_name[MAXPATHLEN];	/* name of the pid file */

int		exit_silent=0;		/* exit silently?	*/

int             timeout_contexts_at_exit=0;     /* timeout contexts when exit? */

struct stat ostb;    /* Logfile attributes, used for -F option */
struct stat nstb;    /* Logfile attributes, used for -F option */

/*
 * definitions for the regular expression matching routines
 *
 * we have an array with submatches and the number of valid submatches
 */
struct re_registers regex_submatches;
struct re_registers regex_notmatches;
int 		regex_submatches_num;

/*
 * new argv and argc for the externals programs we are starting
 */
char		*new_argv[256];	/* build a new argv[] list 	*/
int		new_argc=0;

#ifdef SENDMAIL_FLUSH
/*
 * should we flush the mail-queue?
 */
long		flush_time;

#endif	/* SENDMAIL_FLUSH */



/*
 * process a line from the logfile
 */
void
process_logline()
{
	struct context		*this_context;
	struct rule		*this_rule, *this_rule_next;
	int			check_stop=0;

	if ( logline[0] == '\0' )
		return;
	if ( logline[strlen(logline)-1] == '\n' )
		logline[strlen(logline)-1]='\0';

#ifdef DEBUG
(void) printf("processing logline: %s\n", logline);
#endif
	if ( (logline_context=(struct context_line *)malloc(sizeof(struct context_line)))
		== NULL ) {
		(void) fprintf(stderr, "out of memory processing logling: %s\n",
			logline);
		return;
	}
	if ( (logline_context->content=(char *)malloc(strlen(logline)+1)) == NULL ) {
		(void) fprintf(stderr, "out of memory processing logling: %s\n",
			logline);
		(void) free(logline_context);
		return;
	}
	logline_context->linenumber=logline_num;
	logline_context->timestamp=(long) current_time;
	(void) strcpy(logline_context->content, logline);
	logline_context->link_counter=1;

	/*
	 * first try to store the logline in all matching contexts
	 */
	for ( this_context=all_contexts; this_context != NULL;
		this_context=this_context->next)
		if ( re_search(this_context->match_regex, logline, strlen(logline),
			0, strlen(logline), &regex_submatches) >= 0 ) {
			regex_submatches_num=this_context->match_regex->re_nsub;
			if ( this_context->match_not_regex != NULL ) {
				if ( re_search(this_context->match_not_regex, logline,
					strlen(logline), 0, strlen(logline),
					&regex_notmatches) == -1 ) {
						add_to_context(this_context, logline_context);
				}
			}
			else
				add_to_context(this_context, logline_context);
		}

	/*
	 * now check if there is a matching rule for this logline
	 */
	for ( this_rule=all_rules; this_rule != NULL ; ) {
		/* check if the logline matches the first rgex */
		if ( re_search(this_rule->match_regex, logline, strlen(logline),
			0, strlen(logline), &regex_submatches) >= 0 ) {
			regex_submatches_num=this_rule->match_regex->re_nsub;
			this_rule_next=this_rule->next;
			/* if we do have a not-match regex test it */
			if ( this_rule->match_not_regex != NULL ) {
				if ( re_search(this_rule->match_not_regex, logline,
					strlen(logline), 0, strlen(logline),
					&regex_notmatches) == -1 ) {
						check_stop=1;
						if ( this_rule->do_continue == 0 )
							this_rule_next=NULL;
						process_rule(this_rule);
				}
			}
			else {
				check_stop=1;
				if ( this_rule->do_continue == 0 )
					this_rule_next=NULL;
				process_rule(this_rule);
			}
			if ( check_stop == 1 ) {
				/* check for the stop_regex */
				if ( this_rule->stop_regex != NULL ) {

			if ( re_search(this_rule->stop_regex, logline, strlen(logline),
				0, strlen(logline), &regex_submatches) >= 0 ) {
				/* if we do have a not-match stop-regex test it */
				if ( this_rule->stop_not_regex != NULL ) {
					if ( re_search(this_rule->stop_not_regex,
						logline, strlen(logline), 0,
						strlen(logline), &regex_notmatches) == -1 )
						unlink_rule(this_rule);
				}
				else
					unlink_rule(this_rule);
			}

				}
				check_stop=0;
			}
			this_rule=this_rule_next;
		}
		else
			this_rule=this_rule->next;
	}

	logline_context->link_counter--;
	/*
	 * free the new line if not used in any context
	 */
	destroy_context_line(logline_context);

	return;
}


/*
 * reopen the input logfile
 */
void
logfile_reopen(sig)
	int	sig;
/* ARGSUSED */
{
	if ( logfile_name[0] == '\0' )
		return;
	if ( logfile != NULL )
		(void) fclose(logfile);
	if ( (logfile=fopen(logfile_name, "r")) == NULL ) {
		(void) fprintf(stderr, "error reopening logfile %s\n",
			logfile_name);
		exit(6);
	}
	return;
}

/*
 * Fork and run in the background
 */
int
runasdaemon(lfd)
    int lfd;
{
    int fd = -1;
    int maxfd = -1;

    errno = 0;

    switch (fork()) {
        case -1:
            (void) fprintf(stderr, "could not daemonize: fork(): %s\n", strerror(errno));
            exit(11);
        case 0:
            break;
        default:
            exit(0);
    }

    if ( setsid() < 0 ) {
        (void) fprintf(stderr, "could not daemonize: setsid(): %s\n", strerror(errno));
        exit(6);
    }

    switch (fork()) {
        case -1:
            (void) fprintf(stderr, "could not daemonize: fork(): %s\n", strerror(errno));
            exit(11);
        case 0:
            break;
        default:
            exit(0);
    }

    if ( (fd = open(_PATH_DEVNULL, O_RDWR, 0)) < 0) {
        (void) fprintf(stderr, "could not daemonize: %s\n", strerror(errno));
        exit(11);
    }

    if ( dup2(fd, STDIN_FILENO) == -1 ||
         dup2(fd, STDOUT_FILENO) == -1 ||
         dup2(fd, STDERR_FILENO) == -1 ) {
        (void) fprintf(stderr, "could not daemonize: %s\n", strerror(errno));
        exit(11);
    }

    /* (void) chdir("/"); */

#define MAXFD   1024
    if ( (maxfd = getdtablesize()) < 0)
        maxfd = MAXFD;

    for ( fd=STDERR_FILENO+1; fd<maxfd; fd++ ) {
        if (fd != lfd)
            (void) close(fd);
    }
}

/*
 * print usage information and exit
 */
void
usage(progname)
	char	*progname;
{
	(void) fprintf(stderr,
		"usage: %s [-l startline | -r startregex] [-c configfile] [-d dumpfile] [-p pidfile] [-D] [-f] [-F] [-t] [-e] [logfile]\n",
		progname);
	(void) fprintf(stderr, "This is Logsurfer version %s\n", VERSION );
	exit(1);
}


/*
 * main program
 */
int
main(argc, argv)
	int	argc;
	char	**argv;
{
	char		*progname;	/* name of this program		*/
	int		opt;		/* used for options processing	*/
	extern char	*optarg;	/* option argument		*/
	extern int	optind;		/* option index			*/

	char		cf_filename[MAXPATHLEN]; /* configuration filename */
	long		start_line=0;		/* startline within logfile */
	int             start_at_end=0;         /* start at end of file     */
	struct re_pattern_buffer *start_regex=NULL; /* regex for startline  */
	int		do_follow=0;		/* follow the file?	*/
	int		daemonize=0;		/* run in the background */
    int     write_pidfile=0;    /* flag to write PID to file */
	FILE		*pidfile;		/* write pid info to file */

	char		*logline_buffer;	/* buffer for reading	*/
	int		logline_buffer_size;	/* size of buffer	*/
	int		logline_buffer_pos;	/* position within buffer */

	struct sigaction	signal_info;	/* used by sigaction()	*/


	if ( (progname=strrchr(argv[0], '/')) == NULL )
		progname=argv[0];
	else
		progname++;

	/* setup some defaults and initialize the global pointers */
	(void) strcpy(logfile_name, "");
	(void) strcpy(cf_filename, CONFFILE);
	(void) strcpy(dumpfile_name, DUMPFILE);
	all_rules=NULL;
	all_rules_end=NULL;
	all_contexts=NULL;
	all_contexts_end=NULL;
	next_rule_timeout=LONG_MAX;
	next_context_timeout=LONG_MAX;
	logline_num=0;
	logline=NULL;
    (void) memset(&ostb, 0, sizeof(struct stat));
    (void) memset(&nstb, 0, sizeof(struct stat));

#ifdef SENDMAIL_FLUSH
	flush_time=LONG_MAX;
#endif

	(void) time(&current_time);

#ifdef WARN_ROOT
	if ( getuid() == 0 )
		(void) fprintf(stderr, "warning: %s started as root\n", progname);
#endif

	/*
	 * we want regular expression macthing like egrep
	 */
	re_syntax_options=RE_SYNTAX_POSIX_EGREP;

	/*
	 * get memory to store the submatches for a regex-match
	 */
	regex_submatches.num_regs=RE_NREGS;
	if ( (regex_submatches.start=(regoff_t *)
		malloc(RE_NREGS*sizeof(regoff_t))) == NULL ) {
		(void) fprintf(stderr, "unable to allocate memory for submatches\n");
		exit(99);
	}
	if ( (regex_submatches.end=(regoff_t *)
		malloc(RE_NREGS*sizeof(regoff_t))) == NULL ) {
		(void) fprintf(stderr, "unable to allocate memory for submatches\n");
		exit(99);
	}
	regex_notmatches.num_regs=RE_NREGS;
	if ( (regex_notmatches.start=(regoff_t *)
		malloc(RE_NREGS*sizeof(regoff_t))) == NULL ) {
		(void) fprintf(stderr, "unable to allocate memory for notmatches\n");
		exit(99);
	}
	if ( (regex_notmatches.end=(regoff_t *)
		malloc(RE_NREGS*sizeof(regoff_t))) == NULL ) {
		(void) fprintf(stderr, "unable to allocate memory for notmatches\n");
		exit(99);
	}

	/*
	 * get some memory for the logline_buffer
	 */
	logline_buffer_size=4096;
	logline_buffer_pos=0;
	if ( (logline_buffer=(char *)malloc(logline_buffer_size)) == NULL ) {
		(void) fprintf(stderr, "unable to allocate memory for logline_buffer\n");
		exit(99);
	}

	while ( (opt = getopt(argc, argv, "efFc:d:Dl:p:r:st")) != EOF )
		switch(opt) {
                case 'e':
		        /* start processing at the end of the log file */
		        start_at_end=1;
		        break;
		case 'f':
			/* set follow mode on */
			do_follow=1;
			break;
        case 'F':
            /* set follow obsessively mode on (follow rotated files) */
            do_follow=2;
            break;
		case 'c':
			/* specify another config file */
			(void) strcpy(cf_filename, optarg);
			break;
		case 'd':
			/* specify another dump file */
			(void) strcpy(dumpfile_name, optarg);
			break;
		case 'D':
			/* daemonize: close fd's and run in the background */
            daemonize=1;
			break;
		case 'l':
			/* start processing beginning with given line number */
			if ( start_regex != NULL ) {
				(void) fprintf(stderr, "-l and -r can't be used together\n");
				usage(progname);
			}
            if ( start_at_end != 0 ) {
                (void) fprintf(stderr, "-e and -l can't be used together\n");
                usage(progname);
            }
			start_line=atol(optarg);
			break;
		case 'p':
            write_pidfile=1;
            (void) strcpy(pidfile_name, optarg);
			break;
		case 'r':
			/* start processing at first line that matches regex */
			if ( start_line != 0 ) {
				(void) fprintf(stderr, "-l and -r can't be used together\n");
				usage(progname);
			}
            if ( start_at_end != 0 ) {
                (void) fprintf(stderr, "-e and -r can't be used together\n");
                usage(progname);
            }
			if ( (start_regex=(struct re_pattern_buffer *)
				malloc(sizeof(struct re_pattern_buffer))) == NULL ) {
				(void) fprintf(stderr, "out of memory for start regex %s",
					optarg);
				exit(99);
			}

			start_regex->fastmap=(char *)malloc(256);
			start_regex->translate=(char *) 0;
			start_regex->buffer=NULL;
			start_regex->allocated=0;
			if ( re_compile_pattern(optarg, strlen(optarg), start_regex) != 0 ) {
				(void) fprintf(stderr, "error in start-regex %s\n", optarg);
				exit(5);
			}
			start_regex->regs_allocated=REGS_FIXED;
			break;
		case 's':
			exit_silent=1;
			break;
		case 't':
		        timeout_contexts_at_exit=1;
			break;
		default:
			usage(progname);
			break;
		}

	argc -= optind;
	argv += optind;

	if (argc > 1)
		usage(progname);

	if ( read_config(cf_filename) != 0 ) {
		(void) fprintf(stderr, "error reading configfile %s\n", cf_filename);
		exit(2);
	}

	if ( argc == 0 || !strcmp(argv[0], "-") ) {
        if (daemonize) {
			(void) fprintf(stderr, "cannot use stdin when running as a daemon\n");
			exit(6);
        }
		logfile=stdin;
    }
	else {
		if ( (logfile=fopen(strcpy(logfile_name, argv[0]), "r")) == NULL ) {
			(void) fprintf(stderr, "error opening logfile %s\n",
				logfile_name);
			exit(6);
		}
		if (start_at_end) {
		  fseek(logfile,0,SEEK_END);
		}
        if ( do_follow == 2 ) {
            if ( fstat(fileno(logfile), &ostb) == -1 ) {
                (void) fprintf(stderr, "cannot stat logfile\n");
                exit(6);
            }
        }
	}

    if ( daemonize )
        runasdaemon(fileno(logfile));

    if( write_pidfile ){
        /* write pid to a file */
        if ( (pidfile=fopen(pidfile_name, "w")) == NULL ) {
            (void) fprintf(stderr, "unable to open pidfile %s\n", pidfile_name);
            exit(9);
        }
        (void) fprintf(pidfile, "%ld\n", (long) getpid());
        (void) fclose(pidfile);
    }

	if ( start_line != 0 ) {
		logline_num=start_line-1;
		while ( --start_line > 0 ) {
			if ( (logline=readline(logfile, &logline_buffer,
				&logline_buffer_size, &logline_buffer_pos))
				== NULL ) {
				(void) fprintf(stderr,
					"not enough input lines to skip\n");
				exit(7);
			}
			(void) free(logline);
		}
	}
	else while ( start_regex != NULL ) {
		if ( (logline=readline(logfile, &logline_buffer, &logline_buffer_size,
			&logline_buffer_pos)) == NULL ) {
			(void) fprintf(stderr,
				"unable to find inputline for start_regex\n");
			exit(7);
		}
		logline_num++;
		if ( re_search(start_regex, logline, strlen(logline), 0,
			strlen(logline), &regex_submatches) >= 0 ) {
			regfree(start_regex);
			(void) free(start_regex);
			start_regex=NULL;
			process_logline();
		}
		(void) free(logline);
	}

	signal_info.sa_handler=dump_data;
	(void) sigemptyset(&(signal_info.sa_mask));

#ifdef SA_RESTART
	signal_info.sa_flags=SA_RESTART;
#else
	signal_info.sa_flags=0;
#endif

	if ( sigaction(SIGUSR1, &signal_info, NULL) ) {
		(void) fprintf(stderr, "can't set signal to handle USR1 - abort\n");
		exit(8);
	}
	signal_info.sa_handler=logfile_reopen;
	if ( sigaction(SIGHUP, &signal_info, NULL) ) {
		(void) fprintf(stderr, "can't set signal to handle HUP - abort\n");
		exit(8);
	}
	signal_info.sa_handler=logsurfer_exit;
	if ( sigaction(SIGTERM, &signal_info, NULL) ) {
		(void) fprintf(stderr, "can't set signal to handle TERM - abort\n");
		exit(8);
	}
	if ( sigaction(SIGINT, &signal_info, NULL) ) {
		(void) fprintf(stderr, "can't set signal to handle INT - abort\n");
		exit(8);
	}

	while ( logfile != NULL ) {
		logline=readline(logfile, &logline_buffer, &logline_buffer_size,
			&logline_buffer_pos);
		(void) time(&current_time);
		if ( logline != NULL ) {
			logline_num++;
			process_logline();
			(void) free(logline);
		}
		else if ( do_follow ) {
			(void) sleep(1);
            if ( do_follow == 2 && fileno(logfile) != STDIN_FILENO
                && stat(logfile_name, &nstb) != -1 ) {
                    if ( nstb.st_ino != ostb.st_ino || 
                         nstb.st_dev != ostb.st_dev ||
                         nstb.st_rdev != ostb.st_rdev ||
                         nstb.st_nlink == 0 ) {
                        logfile_reopen(SIGHUP);
                        (void) memcpy(&ostb, &nstb, sizeof(struct stat));
                    }
            }
        }
		else {
			(void) fclose(logfile);
			logfile=NULL;
		}
		regex_submatches_num=0;
		if ( next_rule_timeout < (long) current_time )
			check_rule_timeout();
		if ( next_context_timeout < (long) current_time )
			check_context_timeout();
		check_context_linelimit();
#ifdef SENDMAIL_FLUSH
		if ( flush_time != LONG_MAX ) {
			if ( flush_time < (long) current_time ) {
#ifdef DEBUG
(void) printf("time to flush the sendmail queue now\n");
#endif
				flush_queue();
			}
#ifdef DEBUG
			else
(void) printf("next flush time in %ld seconds\n", flush_time-((long) current_time));
#endif
		}
#endif
		while ( waitpid((pid_t) -1, NULL, WNOHANG) > 0 )
			;
	}
	logsurfer_exit(0);

	/* NOTREACHED */
	return(99);
}

